### PR TITLE
AVX-53924: Add SmartGroup support for S2C and external Group

### PR DESCRIPTION
**Pull request by shinrich**
Adding support for the "s2c" and "external" selectors in the smart_group resource.  Also did some clean up to eliminate some cut-n-paste logic and replace duplicated strings with constants.

Added unit tests which seem to be run by the acceptance tests.

Ran manually locally by augmenting the terraform of one of my end-to-end tests to verify that the new group forms are being created.

### CODE REVIEW

File: goaviatrix/smart_group.go
Comment (stewartbutler) [inline goaviatrix/smart_group.go:60]: I'm guessing you rewrote this block so you could use these constants elsewhere to ensure compatibility?

File: goaviatrix/smart_group.go
Comment (stewartbutler) [inline goaviatrix/smart_group.go]: I know you didn't write this originally but I find naming these variables "or" and "and" (below) to be a bit confusing. Can you change it to `orGroup` and `andGroup` or something, to disambiguate from reserved keywords in other languages?

File: goaviatrix/smart_group.go
Comment (stewartbutler) [inline goaviatrix/smart_group.go]: Since you aren't using a pointer, `data.UUID` should be empty string already, right?

If so, this would trim it down further, but may be confusing. Feel free to disregard.

```suggestion
	err := c.PostAPIContext25(ctx, &data, endpoint, form)
	return data.UUID, err
```

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]: I'm not sure I follow the purpose of this block. Can we get a comment here, explaining it?

I'm also not sure if you fixed a bug or if it is a bad conversion, but previously the error would return `x must be empty when cidr is set` regardless if the map contained `cidr`, `fqdn`, or `site`.

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]: nit: might want to convert these either to `goaviatrix.Whatever` consts or just inline it.

File: goaviatrix/smart_group.go
Comment (stewartbutler) [inline goaviatrix/smart_group.go]: Can we get some doc comments on these? Also this is a pretty awkward name, how does it differ from its sibling?

File: goaviatrix/smart_group.go
Comment (stewartbutler) [inline goaviatrix/smart_group.go]: Doc comment?

File: goaviatrix/utils.go
Comment (stewartbutler) [inline goaviatrix/utils.go:290]: Doc comment + tests?

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]: Can you add some comments for maintainability?

**Review by stewartbutler — CHANGES_REQUESTED** [review]
Looks good but I think we need some additional comments to explain what is going on.

File: aviatrix/resource_aviatrix_smart_group.go
Comment (shinrich) [reply aviatrix/resource_aviatrix_smart_group.go]: Yeah, I fixed the error message.  The first tag that would have been checked is cidr.  More were added over time but the message was not updated. 
Will add a comment explaining the cross field validation that is being performed.

File: aviatrix/resource_aviatrix_smart_group.go
Comment (shinrich) [reply aviatrix/resource_aviatrix_smart_group.go]: Right.  Should tidy up the constants in the error too.

File: goaviatrix/smart_group.go
Comment (shinrich) [reply goaviatrix/smart_group.go:60]: Yes, was trying to remove duplicated strings and eliminate what seemed to be some cut-n-paste code.

File: goaviatrix/smart_group.go
Comment (shinrich) [reply goaviatrix/smart_group.go]: Just adding thing wrappers so we could pass in different booleans to handle the two mapped fields differently depending on whether the resulting map is going to the API or not.  Will add comments and try to get inspired by a nicer name.

File: aviatrix/resource_aviatrix_smart_group.go
Comment (tommy-carver) [reply aviatrix/resource_aviatrix_smart_group.go]: If I'm understanding right, the logic is that you can only choose all tagged selectors or not tagged selectors? Can we make that clear in the error too?

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]:
```suggestion
	// exclusiveFilterKeys are the selector key names that cannot be used with any other filters in a single
	// match_expressions block.
	exclusiveFilterKeys = []string{goaviatrix.ExternalKey, goaviatrix.S2CKey, goaviatrix.CidrKey, goaviatrix.FqdnKey}
```

You should probably also add `ExternalKey` to the docs https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/ce4bf8145e0e3d72e3783cf1038e3a8ee3964864/docs/resources/aviatrix_smart_group.md#required

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]:
```suggestion
	// filterKeys are the selector key names that are associated with a type-based selector. This set of keys should be disjoint from the exclusiveFilterKeys
	filterKeys = []string{goaviatrix.TypeKey, goaviatrix.ResIdKey, goaviatrix.AccountIdKey, goaviatrix.AccountNameKey, goaviatrix.NameKey, goaviatrix.RegionKey, goaviatrix.ZoneKey, goaviatrix.TagsPrefix}
```

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]:
```suggestion
		// Handle the case of the non-tag selector
		if matchingKey, ok := goaviatrix.MapContainsOneOfKeys(selectorInfo, exclusiveFilterKeys); ok {
			// Error if both an exclusive and non-exclusive filter selector are present together
			if matchingType, okType := goaviatrix.MapContainsOneOfKeys(selectorInfo, filterKeys); okType {
				return nil, fmt.Errorf("%q must be empty when %q is set", matchingType, matchingKey)
```

Do you also need a condition here for if two exclusive keys are used? E.g. `cidr` and `fqdn`?

File: aviatrix/resource_aviatrix_smart_group.go
Comment (stewartbutler) [inline aviatrix/resource_aviatrix_smart_group.go]:
```suggestion
			return nil, fmt.Errorf("%q is required when %q are all empty", goaviatrix.TypeKey, strings.Join(exclusiveFilterKeys, ", "))
```

File: docs/resources/aviatrix_smart_group.md
Comment (tommy-carver) [inline docs/resources/aviatrix_smart_group.md:56]: Can we add s2c and external as Optional arguments in the argument reference below as well?

Comment (tommy-carver) [reply docs/resources/aviatrix_smart_group.md:56]: Also what is ext_args and does that need to be added as well?

**Review by tommy-carver — COMMENTED** [review]
Just some nits

**Review by stewartbutler — COMMENTED** [review]
Few more nits (mostly var name changes)

File: aviatrix/resource_aviatrix_smart_group.go
Comment (shinrich) [reply aviatrix/resource_aviatrix_smart_group.go]: As we discussed. I'm just going to pull this logic.  The API is already doing this and additional validation.  No need to duplicate here.

**Review by stewartbutler — DISMISSED** [review]
LGTM, but needs to wait for @rpanzer-aviatrix's review as well.

**Review by tommy-carver — DISMISSED** [review]
LGTM, but wait for @rpanzer-aviatrix  approval as well

File: goaviatrix/smart_group.go
Comment (rpanzer-aviatrix) [inline goaviatrix/smart_group.go:76]: It seems like a typo sneaked in:
```suggestion
	K8sServiceKey   = "k8s_service"
	K8sPodNameKey   = "k8s_pod"
```

File: aviatrix/resource_aviatrix_smart_group_test.go
Comment (rpanzer-aviatrix) [inline aviatrix/resource_aviatrix_smart_group_test.go]: I wonder how this should work. `Config: testAccSmartGroupK8s()` creates `aviatrix_smart_group.k8s`.
But the `Check` tests for a resource named `aviatrix_smart_group.s2c`.
Hence the test also fails for me:
```
=== RUN   TestAccAviatrixSmartGroup_s2c
time="2024-08-20T10:46:06+02:00" level=info msg="Getting API token..."
time="2024-08-20T10:46:07+02:00" level=info msg="Parsed Aviatrix login: admin"
time="2024-08-20T10:46:08+02:00" level=info msg="Getting API token..."
time="2024-08-20T10:46:08+02:00" level=info msg="Parsed Aviatrix login: admin"
time="2024-08-20T10:46:09+02:00" level=info msg="Getting API token..."
time="2024-08-20T10:46:09+02:00" level=info msg="Parsed Aviatrix login: admin"
    resource_aviatrix_smart_group_test.go:211: Step 1/2 error: Check failed: Check 1/5 error: no Smart Group resource found: aviatrix_smart_group.s2c
time="2024-08-20T10:46:10+02:00" level=info msg="Getting API token..."
time="2024-08-20T10:46:11+02:00" level=info msg="Parsed Aviatrix login: admin"
time="2024-08-20T10:46:11+02:00" level=info msg="Getting API token..."
time="2024-08-20T10:46:12+02:00" level=info msg="Parsed Aviatrix login: admin"
--- FAIL: TestAccAviatrixSmartGroup_s2c (6.73s)
```
Ideally it should even fail since I probably don't have the correct version of the controller running.

File: aviatrix/resource_aviatrix_smart_group_test.go
Comment (rpanzer-aviatrix) [inline aviatrix/resource_aviatrix_smart_group_test.go]: Same here as above.

File: aviatrix/resource_aviatrix_smart_group_test.go
Comment (rpanzer-aviatrix) [inline aviatrix/resource_aviatrix_smart_group_test.go]: Same here as above.

**Review by rpanzer-aviatrix — CHANGES_REQUESTED** [review]
Requesting changes because the typos in the const values of `K8sServiceKey` and `K8sPodNameKey` break the existing Kubernetes functionality.

File: aviatrix/resource_aviatrix_smart_group_test.go
Comment (shinrich) [reply aviatrix/resource_aviatrix_smart_group_test.go]: Interesting.  I agree it should fail, but it does succeed in my dev environment.  Hmm, just added the -v arguments.  Looks like I'm triggering some skip logic since TF_ACC is not set.

**Review by rpanzer-aviatrix — APPROVED** [review]

File: goaviatrix/smart_group.go
Comment (rpanzer-aviatrix) [reply goaviatrix/smart_group.go:76]: Weird, I thought this was fixed. 
Seems like it went into master.

**Pull request merged**
